### ISSUE TITLE
Nested components more than 2 levels deep fix in newer versions of Strapi

### ISSUE THREAD

**Issue by kevinvugts** [strapi/strapi#11319]
### Please describe your feature request

- [ ] **I have created my request on the Product Board before I submitted this issue**
- [x ] **I have looked at all the other requests on the Product Board before I submitted this issue**

### Summary

Quick summary what's this feature request about.

Recently you guys enforced the depth restriction to a maximum of 2 levels eep on data loading to avoid over fetching data et avoid infinite loops. As discussed in this #9462 @alexandrebodin created a PR fix for this in this[ pull request](https://github.com/strapi/strapi/pull/9464).

### Why is it needed?

Currently while building websites with Strapi as a back-end it annoys us as developers that we cannot use more than 2 levels deep of nesting components in the newer versions of Strapi. Take for example a Hero component including a set of cards, and a set of cards also includes buttons which link to different sections in your website. This is simply not possible with newer versions in Strapi.

### Suggested solution(s)

As discussed with @derrickmehaffy on Discord I created this new issue because the old aged issue isn't getting much attention really because of the more important V4 stuff which is in the pipeline. Is this issue still present in the V4 of Strapi? 

### Related issue(s)/PR(s)

Issue: https://github.com/strapi/strapi/issues/9462
PR: https://github.com/strapi/strapi/pull/9464

PS: @derrickmehaffy if necessary, can you add more context to it? 

Thanks in advance! 👍

**Comment by AtticusCMBM** [comment]
I also would really like this feature. Just picked up Strapi and I ran into this issue in the first day. Thank you coder people,

**Comment by anthonytison** [comment]
+1 
It's really frustrating being restricted on this feature. I've been blocked on my development many times by this restriction.

**Comment by Kinbaum** [comment]
Bump. I feel like this shouldn't be a restriction, especially given the new population option. I find it pretty restrictive even for the most basic structure to have this limitation. If this is a UI/UX concern for authors, then I'd suggest providing components for developers to specifically tackle this problem so we can improve these experiences ourselves. Components likes Tabs and Accordions would allow us to logically group things together / just make a better experience for authors.

**Comment by Jason-Kennemer** [comment]
In version 4.0.0, nested components are not available at all in the admin. I tried to re-create one of the Strapi Next.js examples, and I cannot see or use components that contain components in them. This makes the product unusable.

**Comment by MarcosGin** [comment]
Bug related with this https://github.com/strapi/strapi/issues/11955.

we need more than 2 nested component.

this is hardcoded in the entity manager

<img width="776" alt="Screen Shot 2021-12-22 at 19 34 32" src="https://user-images.githubusercontent.com/26859121/147162675-b8c84837-3e14-4e70-a90f-ecff735e16ae.png">

It is true that at the ui level you cannot add more than two components, which I think is inefficient.
But if the developer creates these configurations manually, with the json files, he can perform these interactions at a higher level.

Here is when the admin fails, because it does not bring the properties to that level. I think we need to improve this and make it free for the developer.

**Comment by Kinbaum** [comment]
@derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?

**Comment by kevinvugts** [comment]
> @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?

me to @derrickmehaffy  👍

**Comment by tilman** [comment]
I have described a possible solution in: https://github.com/strapi/strapi/issues/12126#issuecomment-1006805837
You can use patch-package to disable the check in the frontend.

**Comment by Kinbaum** [comment]
@tilman While the UI now display the component, when I attempt to save the model in the UI, it throws a bad request with the following:

```
card already is a nested component. You cannot have more than one level of nesting inside your components."
```

The structure in my example is:
card group > card > link

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?



> > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> 
> me to @derrickmehaffy +1

Yes it is being discussed, nothing to share at this time.

**Comment by tilman** [comment]
@Kinbaum you are right! Seems like I forgot to add one of the files to the patch-file. I have updated the diff in my comment (https://github.com/strapi/strapi/issues/12126#issuecomment-1006805837) and added the patch for the server controller as well. Let me know if it still not works.

**Comment by Kinbaum** [comment]
@tilman Nice! That bypasses the problem for the time being. With this type of setup where you have the composability with components, the queries sort of become tricky. There needs to be a way to just get everything, since you wont really know what to populate since a content author can change the structures / introduce new content at any time.

**Comment by fperaltalitebox** [comment]
This really sad :(

**Comment by Kinbaum** [comment]
As part of this discussion, I believe that you should also have the ability to nest a dynamic zone within another dynamic zone. This type of capability will enable authors to build flexible layouts within the interface (Eg. a Grid component with a multi-column design). Again though, you'd have to have an "auto-populating" mechanism for components / relations especially when using frameworks like Next.js where you have SSG / ISR. You won't know what the REST / GraphQL queries should be because an author can change whatever content they want at any time.

**Comment by pshemekm** [comment]
I also ran into this issue on day one:

`Cta -> Button -> Link`

I'd vote for a configuration option to skip this validation or to choose the allowed level.

**Comment by Jimmylet** [comment]
@alexandrebodin any news?
I ran into the problem on my first day of testing with Strapi and it's really frustrating.
We cannot, as professionals who deliver products, do without this feature. Customers wouldn't understand that.
We are looking for a "headless cms" to develop our agency. Strapi is very promising but not having this type of functionality makes it unusable.
Using relationships to do this is not a good solution.

**Comment by kevinvugts** [comment]
> > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> 
> > > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> > 
> > 
> > me to @derrickmehaffy +1
> 
> Yes it is being discussed, nothing to share at this time.

Jo Derrick,

As this issue becomes more and more a relevant topic and enhancement / rollback of strapi v4. Could you please elaborate more on the next steps?

Is there still nothing to be shared?

**Comment by derrickmehaffy** [comment]
> > > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> > 
> > 
> > > > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> > > 
> > > 
> > > me to @derrickmehaffy +1
> > 
> > 
> > Yes it is being discussed, nothing to share at this time.
> 
> Jo Derrick,
> 
> As this issue becomes more and more a relevant topic and enhancement / rollback of strapi v4. Could you please elaborate more on the next steps?
> 
> Is there still nothing to be shared?

The team is quite busy with higher priority stuff at the moment, the only answer I have right now is that it's not in our roadmap for Q1 and it hasn't been discussed for Q2.

(I'm not a decision maker on this topic, just the messenger) 

CC @yanniskadiri

**Comment by kevinvugts** [comment]
> > > > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> > > 
> > > 
> > > > > @derrickmehaffy I was wondering if there have been any developments/talks with regards to this issue recently?
> > > > 
> > > > 
> > > > me to @derrickmehaffy +1
> > > 
> > > 
> > > Yes it is being discussed, nothing to share at this time.
> > 
> > 
> > Jo Derrick,
> > As this issue becomes more and more a relevant topic and enhancement / rollback of strapi v4. Could you please elaborate more on the next steps?
> > Is there still nothing to be shared?
> 
> The team is quite busy with higher priority stuff at the moment, the only answer I have right now is that it's not in our roadmap for Q1 and it hasn't been discussed for Q2.
> 
> (I'm not a decision maker on this topic, just the messenger)
> 
> CC @yanniskadiri

Cheers bud for your answer.

**Comment by lukasz-mical** [comment]
@kevinvugts  @Jimmylet Until the Strapi team fixes it, you can use the workaround that I am using now.
1. Install patch-package `yarn add patch-package`
2. Add script to your packages.json `"postinstall": "patch-package"`
3. Add file `patches/@strapi+plugin-content-manager+4.0.4.patch`
4. Paste code
```patch
diff --git a/node_modules/@strapi/plugin-content-manager/server/services/entity-manager.js b/node_modules/@strapi/plugin-content-manager/server/services/entity-manager.js
index df4b28e..c0910f0 100644
--- a/node_modules/@strapi/plugin-content-manager/server/services/entity-manager.js
+++ b/node_modules/@strapi/plugin-content-manager/server/services/entity-manager.js
@@ -1,6 +1,7 @@
 'use strict';
 
 const { assoc, has, prop, omit } = require('lodash/fp');
+const merge = require('lodash/merge');
 const strapiUtils = require('@strapi/utils');
 const { ApplicationError } = require('@strapi/utils').errors;
 
@@ -39,13 +40,15 @@ const findCreatorRoles = entity => {
   return [];
 };
 
+const deepMerge = (oldData, newData) => oldData ? merge(oldData, newData) : newData;
+
 // TODO: define when we use this one vs basic populate
 const getDeepPopulate = (uid, populate, depth = 0) => {
   if (populate) {
     return populate;
   }
 
-  if (depth > 2) {
+  if (depth > 4) {
     return {};
   }
 
@@ -73,7 +76,7 @@ const getDeepPopulate = (uid, populate, depth = 0) => {
     if (attribute.type === 'dynamiczone') {
       populateAcc[attributeName] = {
         populate: (attribute.components || []).reduce((acc, componentUID) => {
-          return Object.assign(acc, getDeepPopulate(componentUID, null, depth + 1));
+          return deepMerge(acc, getDeepPopulate(componentUID, null, depth + 1));
         }, {}),
       };
     }

```
5. Add file `patches/@strapi+plugin-content-type-builder+4.0.4.patch`
6. Paste code
```patch
diff --git a/node_modules/@strapi/plugin-content-type-builder/admin/src/components/SelectComponent/index.js b/node_modules/@strapi/plugin-content-type-builder/admin/src/components/SelectComponent/index.js
index 0c692d8..5d96d05 100644
--- a/node_modules/@strapi/plugin-content-type-builder/admin/src/components/SelectComponent/index.js
+++ b/node_modules/@strapi/plugin-content-type-builder/admin/src/components/SelectComponent/index.js
@@ -47,11 +47,11 @@ const SelectComponent = ({
     return [...acc, ...compos];
   }, []);
 
-  if (isAddingAComponentToAnotherComponent) {
-    options = options.filter(option => {
-      return !componentsThatHaveOtherComponentInTheirAttributes.includes(option.uid);
-    });
-  }
+  // if (isAddingAComponentToAnotherComponent) {
+  //   options = options.filter(option => {
+  //     return !componentsThatHaveOtherComponentInTheirAttributes.includes(option.uid);
+  //   });
+  // }
 
   if (isTargetAComponent) {
     options = options.filter(option => {
diff --git a/node_modules/@strapi/plugin-content-type-builder/server/controllers/validation/types.js b/node_modules/@strapi/plugin-content-type-builder/server/controllers/validation/types.js
index 735f4ea..faab13f 100644
--- a/node_modules/@strapi/plugin-content-type-builder/server/controllers/validation/types.js
+++ b/node_modules/@strapi/plugin-content-type-builder/server/controllers/validation/types.js
@@ -237,12 +237,12 @@ const getTypeShape = (attribute, { modelType, attributes } = {}) => {
               const targetCompo = strapi.components[compoUID];
               if (!targetCompo) return true; // ignore this error as it will fail beforehand
 
-              if (modelType === modelTypes.COMPONENT && hasComponent(targetCompo)) {
-                return this.createError({
-                  path: this.path,
-                  message: `${targetCompo.modelName} already is a nested component. You cannot have more than one level of nesting inside your components.`,
-                });
-              }
+              // if (modelType === modelTypes.COMPONENT && hasComponent(targetCompo)) {
+              //   return this.createError({
+              //     path: this.path,
+              //     message: `${targetCompo.modelName} already is a nested component. You cannot have more than one level of nesting inside your components.`,
+              //   });
+              // }
               return true;
             },
           })
```
7. Run install `yarn install`
8. Remove `.cache` and `build` folders
9. Build admin `yarn build`

This solution works for me

**Comment by blushbasics** [comment]
First day using strapi and already regretting my decision based on this thread ☹️

![adminError](https://user-images.githubusercontent.com/74014574/153088871-0cc35d85-b7ba-474b-ad7f-2d3a10ca30a2.png)

![error](https://user-images.githubusercontent.com/61177770/153088055-4bbeb0ec-b425-4f5d-9dc7-77f3a2790ce7.png)

This is my content structure. Not sure why this isn't possible?

![content](https://user-images.githubusercontent.com/61177770/153088269-a73bb034-64ee-4560-a22b-8e7b5a24310f.png)

**Comment by gwynhokkers** [comment]
> @kevinvugts @Jimmylet Until the Strapi team fixes it, you can use the workaround that I am using now.

> This solution works for me
@lukasz-mical 
Hey, got this working on 4.0.4, which is great, many, many thanks. What do we need to do, to update this to 4.1.3?

**Comment by derrickmehaffy** [comment]
Hey all, we are currently moving many feature requests to our new feedback and feature request website to help clean up the GitHub issues and make it easier for us to review bug reports and fix them.

I have moved this feature to the following URL: https://feedback.strapi.io/customization/p/allow-nesting-of-components-more-than-2-levels-deep
If you are interested in this feature please feel free to go there and add more comments and/or upvote it there.

For now I will close this and lock it so that all new information goes into our new feedback website.
Thanks!
